client/consensus/common/src/level_monitor.rs

```diff
@@ -0,0 +1,325 @@
+	///
+	/// Level limits are not enforced during this phase.
+	fn restore(&mut self) {
+		let mut counter_max = Zero::zero();
+		let info = self.backend.blockchain().info();
+
+		self.import_counter = info.finalized_number;
+		self.update(info.finalized_number, info.finalized_hash);
+
+		for leaf in self.backend.blockchain().leaves().unwrap_or_default() {
+			let route =
+				sp_blockchain::tree_route(self.backend.blockchain(), info.finalized_hash, leaf)
+					.expect("Route from finalized to leaf should be available; qed");
+			if !route.retracted().is_empty() {
+				continue
+			}
+			route.enacted().iter().for_each(|elem| {
+				if !self.freshness.contains_key(&elem.hash) {
+					// Use the block height value as the freshness.
+					self.import_counter = elem.number;
+					self.update(elem.number, elem.hash);
+				}
+			});
+			if self.import_counter > counter_max {
+				counter_max = self.import_counter;
+			}
+		}
+
+		self.import_counter = counter_max;
+		self.lowest_level = self.backend.blockchain().info().finalized_number;
+	}
+
+	/// Check and enforce the limit bound at the given height.
+	///
+	/// In practice this will enforce the given height in having a number of blocks less than
+	/// the limit passed to the constructor.
+	///
+	/// If the given level is found to have a number of blocks greater than or equal the limit
+	/// then the limit is enforced by chosing one (or more) blocks to remove.
+	///
+	/// The removal strategy is driven by the block freshness.
+	///
+	/// A block freshness is determined by the most recent leaf freshness descending from the block
+	/// itself. In other words its freshness is equal to its more "fresh" descendant.
+	///
+	/// The least "fresh" blocks are eventually removed.
+	pub fn enforce_limit(&mut self, number: NumberFor<Block>) {
+		if self.import_counter == Zero::zero() {
+			// First call detected, restore using the backend persistent information.
+			self.restore();
+		}
+
+		let level_len = self.levels.get(&number).map(|l| l.len()).unwrap_or_default();
+		if level_len < self.level_limit {
+			return
+		}
+
+		log::debug!(
+			target: "parachain",
+			"Detected leaves overflow at height {}, removing obsolete blocks",
+			number,
+		);
+
+		let mut leaves = self.backend.blockchain().leaves().unwrap_or_default();
+		// Sort leaves by freshness (less fresh first).
+		leaves.sort_unstable_by(|a, b| self.freshness.get(a).cmp(&self.freshness.get(b)));
+
+		// This may not be the most efficient way to remove **multiple** entries, but is the easy
+		// one :-). Should be considered that in "normal" conditions the number of blocks to remove
+		// is 0 or 1, it is not worth to complicate the code too much. One condition that may
+		// trigger multiple removals (2+) is if we restart the node using an existing db and a
+		// smaller limit wrt the one previously used.
+		let remove_count = level_len - self.level_limit + 1;
+		(0..remove_count).all(|_| {
```

Comment:
Within the iteration the only thing that could fail is `find_target()`.
If it fails once then it will fail for all the other attempts as well.
I've used `all` to stop as soon as it fails the first time